### PR TITLE
ELEX-2763: Estimandizer update

### PR DESCRIPTION
**Pull request by dmnapolitano**
## Description

The changes in this PR refine the changes made in PR https://github.com/washingtonpost/elex-live-model/pull/59.  Hopefully the code and logic are much cleaner now.  Thanks to @lennybronner for making most of these changes lol and the helpful feedback 😄 🙌🏻 

Note that, as a result of these changes, a change will need to be made to the test bed; specifically, in `modeltestbed/redo.py`, `include_results_estimand=True` will need to be added to the constructor.  Previously, I was trying to avoid having to modify the test bed, but that lead to some ugliness 🙈. But this change is extremely minimal and the rest of the code is much cleaner now, so IMO, it's worth it.

## Jira Ticket

ELEX-2763

## Test Steps

`tox` and you can also modify the test bed per the above (until the test bed itself is modified, of course).  Thanks!

### CODE REVIEW

**Comment by dmnapolitano** [comment]
Hi @lennybronner , I'm stuck and need help with the integration tests ☹️ 

The one failing is this one:

```
elexmodel 2021-11-02_VA_G --estimands=dem --office_id=G --geographic_unit_type=precinct --percent_reporting 60 --historical --aggregates=county_classification
```

The issue is that `baseline_dem` is present but not `results_dem`, so the `Estimandizer` is trying (reasonably) to estimandize a `results_dem` and failing due to a lack of `dem` estimandizer.  What do we want to do in this situation?  Do we want to copy `baseline_dem` to `results_dem`?  This is being called from `MockLiveDataHandler` 🤔

File: src/elexmodel/handlers/data/Estimandizer.py [hunk 1]
```diff
@@ -14,22 +14,14 @@ class Estimandizer:
...
-                    # will raise a KeyError if a function with the same name as `estimand` doesn't exist
-                    data_df = globals()[estimand](data_df)
-                if target_col == baseline_col:
-                    data_df[results_col] = data_df[baseline_col].copy()
+
+            if results_col not in data_df.columns:
+                # will raise a KeyError if a function with the same name as `estimand` doesn't exist
+                data_df = globals()[estimand](data_df, RESULTS_PREFIX)
 
             if historical:
```
Comment (lennybronner) [inline src/elexmodel/handlers/data/Estimandizer.py:26]: why is this part necessary?

File: tests/handlers/test_estimandizer.py [hunk 2]
```diff
@@ -27,24 +26,25 @@ def test_check_and_create_estimands_historical(va_governor_county_data):
     estimandizer = Estimandizer()
     (output_df, result_columns) = estimandizer.check_and_create_estimands(va_data_copy, estimands, True)
 
-    assert "baseline_party_vote_share_dem" in output_df.columns
     assert "results_party_vote_share_dem" in output_df.columns
     assert result_columns == ["results_party_vote_share_dem"]
+    assert output_df["results_party_vote_share_dem"].isnull().all()
```
Comment (lennybronner) [inline tests/handlers/test_estimandizer.py:31]: what is this testing?

**Review by lennybronner — COMMENTED** [review]
This looks v good! Though atm the integration tests are failing

**Comment by dmnapolitano** [comment]
> This looks v good! Though atm the integration tests are failing. Also when I run a historical election all subunits are unexpected.

Thanks!  Sorry about the failing integration tests.  I have a question about that in a comment for you (above) because I'm confused about how we're expecting that test to work 🤔 

In regards to the unexpected subunits, can you provide an example for that that I can look into? 🤔

File: src/elexmodel/client.py [hunk 3]
```diff
@@ -391,6 +391,23 @@ def _format_historical_current_data(
         """
         Formats data for historical model run
         """
+
+        """
```
[the remark this answers is not in the document]
Comment (dmnapolitano) [reply src/elexmodel/client.py:395]: Sure!  I find it helpful so I vote for keeping it 🗳️  But sorry, what are normal comments? 🤔

File: src/elexmodel/handlers/data/Estimandizer.py [hunk 4]
```diff
@@ -45,44 +37,45 @@ def check_and_create_estimands(self, data_df, estimands, historical, current_dat
...
 
 
-def party_vote_share_dem(data_df):
-    # should only happen when we're replaying an election
-    if f"{BASELINE_PREFIX}dem" not in data_df.columns and f"{BASELINE_PREFIX}turnout" not in data_df.columns:
-        data_df[f"{RESULTS_PREFIX}party_vote_share_dem"] = (
-            data_df[f"{RESULTS_PREFIX}dem"] / data_df[f"{RESULTS_PREFIX}turnout"]
-        )
+def party_vote_share_dem(data_df, col_prefix):
+    if f"{col_prefix}dem" in data_df.columns and f"{col_prefix}turnout" in data_df.columns:
```
[the remark this answers is not in the document]
Comment (dmnapolitano) [reply src/elexmodel/handlers/data/Estimandizer.py:70]: I was testing this out with the test bed and the `2021-11-02_VA_G --office_id Y --geographic_unit_type "precinct-district" --no_ap` command we've been talking about.  In that data set, there are only "dem" and "total" columns, nothing like `results_dem` or `baseline_dem` like we'd need here.  So I had to add this check and use "dem" and "total" columns as a fall-back 🤔

File: src/elexmodel/handlers/data/Estimandizer.py (hunk 1 above)
Comment (dmnapolitano) [reply src/elexmodel/handlers/data/Estimandizer.py:26]: I migrated code from `MockLiveDataHandler` in the previous PR that did this.  I didn't want anything to break.  Hmmmm 🤔

File: tests/handlers/test_estimandizer.py (hunk 2 above)
Comment (dmnapolitano) [reply tests/handlers/test_estimandizer.py:31]: When we're working with a historical election and calling (the soon-to-be-renamed) `check_and_create_estimands()`, `results_estimand` columns are being set to `nan`.  This is to make sure that happens for this new custom estimand we've added 🤔

File: src/elexmodel/handlers/data/Estimandizer.py [hunk 5]
```diff
@@ -14,22 +14,14 @@ class Estimandizer:
     Estimandizer. Generate estimands explicitly.
     """
 
-    def check_and_create_estimands(self, data_df, estimands, historical, current_data=False):
+    def check_and_create_estimands(self, data_df, estimands, historical):
```
[the remark this answers is not in the document]
Comment (dmnapolitano) [reply src/elexmodel/handlers/data/Estimandizer.py]: Done! 🎉

File: src/elexmodel/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lennybronner) [reply src/elexmodel/client.py:395]: haha oops sorry, comments that start with `#` rather than the docstring ones

File: src/elexmodel/handlers/data/Estimandizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lennybronner) [reply src/elexmodel/handlers/data/Estimandizer.py:70]: ooh yes, I know what you mean! I think it would be better to deal with that in the testbed. At some point I changed the name of the column that was being stored in s3 and I think the model should assume that whatever we pass to it starts with `results_dem` etc.

File: src/elexmodel/handlers/data/Estimandizer.py (hunk 1 above)
Comment (lennybronner) [reply src/elexmodel/handlers/data/Estimandizer.py:26]: So I think it makes sense in the `MockLiveData` case when historical is true (since we are overwriting those results anyway), but it doesn't make sense in the `PreprocessedData` if we call that function on line 61 (since we need the results for the historical run). I think it might be worth just getting rid  of the `nan` generally? We don't need those results in the first case, but they aren't actually a problem, I think?

**Comment by lennybronner** [comment]
> > This looks v good! Though atm the integration tests are failing. Also when I run a historical election all subunits are unexpected.
> 
> Thanks! Sorry about the failing integration tests. I have a question about that in a comment for you (above) because I'm confused about how we're expecting that test to work 🤔
> 
> In regards to the unexpected subunits, can you provide an example for that that I can look into? 🤔

oh oops, sorry I missed that. How was this test passing earlier? 🤔 

Here is an example of all units being unexpected:
```
elexmodel 2020-11-03_USA_G --office_id=P --estimands=party_vote_share_dem --geographic_unit_type=county --pi_method=nonparametric --percent_reporting=30 --aggregates=postal_code --historical
```

File: src/elexmodel/handlers/data/Estimandizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmnapolitano) [reply src/elexmodel/handlers/data/Estimandizer.py:70]: Yeah, I agree.  As I was working on this I was wondering how often we'd be updating estimandizers to accommodate new column names lol 😅  I'll take this out and add some logic in the test bed.  Thanks 😄

File: src/elexmodel/handlers/data/Estimandizer.py (hunk 1 above)
Comment (dmnapolitano) [reply src/elexmodel/handlers/data/Estimandizer.py:26]: I see, hmmmm...ok, I'll take it out and see what breaks 😄
